thrall/app/lib/kinesis/MessageTranslator.scala

```diff
@@ -0,0 +1,71 @@
+      }
+      case UpdateImageExports => (updateMessage.id, updateMessage.crops) match {
+        case (Some(id), Some(crops)) => Right(UpdateImageExportsMessage(id, updateMessage.lastModified, crops))
+        case _ => Left(MissingFieldsException(updateMessage.subject))
+      }
+      case UpdateImageUserMetadata => (updateMessage.id, updateMessage.edits) match {
+        case (Some(id), Some(edits)) => Right(UpdateImageUserMetadataMessage(id, updateMessage.lastModified, edits))
+        case _ => Left(MissingFieldsException(updateMessage.subject))
+      }
+      case UpdateImageUsages => (updateMessage.id, updateMessage.usageNotice) match {
+        case (Some(id), Some(usageNotice)) => Right(UpdateImageUsagesMessage(id, updateMessage.lastModified, usageNotice))
+        case _ => Left(MissingFieldsException(updateMessage.subject))
+      }
+      case ReplaceImageLeases => (updateMessage.id, updateMessage.leases) match {
+        case (Some(id), Some(leases)) => Right(ReplaceImageLeasesMessage(id, updateMessage.lastModified, leases))
+        case _ => Left(MissingFieldsException(updateMessage.subject))
+      }
+      case AddImageLease => (updateMessage.id, updateMessage.mediaLease) match {
+        case (Some(id), Some(mediaLease)) => Right(AddImageLeaseMessage(id, updateMessage.lastModified, mediaLease))
+        case _ => Left(MissingFieldsException(updateMessage.subject))
+      }
+      case RemoveImageLease => (updateMessage.id, updateMessage.leaseId) match {
+        case (Some(id), Some(leaseId)) => Right(RemoveImageLeaseMessage(id, updateMessage.lastModified, leaseId))
+        case _ => Left(MissingFieldsException(updateMessage.subject))
+      }
+      case SetImageCollections => (updateMessage.id, updateMessage.collections) match {
+        case (Some(id), Some(collections)) => Right(SetImageCollectionsMessage(id, updateMessage.lastModified, collections))
+        case _ => Left(MissingFieldsException(updateMessage.subject))
+      }
+      case DeleteUsages => (updateMessage.id) match {
+        case Some(id) => Right(DeleteUsagesMessage(id, updateMessage.lastModified))
+        case _ => Left(MissingFieldsException(updateMessage.subject))
+      }
+
+      case UpdateImageSyndicationMetadata => (updateMessage.id, updateMessage.syndicationRights) match {
+        case (Some(id), maybeSyndicationMetadata) => Right(UpdateImageSyndicationMetadataMessage(id, updateMessage.lastModified, maybeSyndicationMetadata))
+        case _ => Left(MissingFieldsException(updateMessage.subject))
+      }
+      case UpdateImagePhotoshootMetadata => (updateMessage.id, updateMessage.edits) match {
+        case (Some(id), Some(edits)) => Right(UpdateImagePhotoshootMetadataMessage(id, updateMessage.lastModified, edits))
+        case _ => Left(MissingFieldsException(updateMessage.subject))
+      }
+      case _ => Left(ProcessorNotFoundException(updateMessage.subject))
+    }
```

Comment:
Personally I have a pathological dislike of nested pattern matching and prefer something like this (granted this then has a lot of underscores, so I understand the motivation):
```suggestion
    updateMessage match {
      case UpdateMessage(Image | ReingestImage | UpdateImage, Some(image), Some(id), _, _, lastModified, _, _, _, _, _, _, _) =>
        Right(ImageMessage(id, lastModified, image))
      case UpdateMessage(DeleteImage, _, Some(id), _, _, lastModified, _, _, _, _, _, _, _) =>
        Right(DeleteImageMessage(id, lastModified))
      case UpdateMessage(DeleteImageExports, _, Some(id), _, _, lastModified, _, _, _, _, _, _, _) =>
        Right(DeleteImageExportsMessage(id, lastModified))
      case UpdateMessage(UpdateImageExports, _, Some(id), _, _, lastModified, _, _, Some(crops), _, _, _, _) =>
        Right(UpdateImageExportsMessage(id, lastModified, crops))
      case UpdateMessage(UpdateImageUserMetadata, _, Some(id), _, Some(edits), lastModified, _, _, _, _, _, _, _) =>
        Right(UpdateImageUserMetadataMessage(id, lastModified, edits))
      case UpdateMessage(UpdateImageUsages, _, Some(id), Some(usageNotice), _, lastModified, _, _, _, _, _, _, _) =>
        Right(UpdateImageUsagesMessage(id, lastModified, usageNotice))
      case UpdateMessage(UpdateImageUsages, _, Some(id), _, _, lastModified, _, _, _, _, Some(leases), _, _) =>
        Right(ReplaceImageLeasesMessage(id, lastModified, leases))
      case UpdateMessage(AddImageLease, _, Some(id), _, _, lastModified, _, _, _, Some(mediaLease), _, _, _) =>
        Right(AddImageLeaseMessage(id, lastModified, mediaLease))
      case UpdateMessage(RemoveImageLease, _, Some(id), _, _, lastModified, _, Some(leaseId), _, _, _, _, _) =>
        Right(RemoveImageLeaseMessage(id, lastModified, leaseId))
      case UpdateMessage(SetImageCollections, _, Some(id), _, _, lastModified, Some(collections), _, _, _, _, _, _) =>
        Right(SetImageCollectionsMessage(id, lastModified, collections))
      case UpdateMessage(DeleteUsages, _, Some(id), _, _, lastModified, _, _, _, _, _, _, _) =>
        Right(DeleteUsagesMessage(id, lastModified))
      case UpdateMessage(UpdateImageSyndicationMetadata, _, Some(id), _, _, lastModified, _, _, _, _, _, maybeSyndicationMetadata, _) =>
        Right(UpdateImageSyndicationMetadataMessage(id, lastModified, maybeSyndicationMetadata))
      case UpdateMessage(UpdateImagePhotoshootMetadata, _, Some(id), _, Some(edits), lastModified, _, _, _, _, _, _, _) =>
        Right(UpdateImagePhotoshootMetadataMessage(id, updateMessage.lastModified, edits))
      case UpdateMessage(subject, _, _, _, _, _, _, _, _, _, _, _, _) if validSubjects.contains(subject) =>
        Left(MissingFieldsException(updateMessage.subject))
      case UpdateMessage(subject, _, _, _, _, _, _, _, _, _, _, _, _) =>
        Left(ProcessorNotFoundException(updateMessage.subject))
    }
```

Reply:
You'll need this in the MessageSubjects trait if you take the above (shame we don't have enumeratum in this project yet...):
```
  val validSubjects = Set(
    Image,
    ReingestImage,
    DeleteImage,
    UpdateImage,
    DeleteImageExports,
    UpdateImageExports,
    UpdateImageUserMetadata,
    UpdateImageUsages,
    ReplaceImageLeases,
    AddImageLease,
    RemoveImageLease,
    SetImageCollections,
    DeleteUsages,
    UpdateImageSyndicationMetadata,
    UpdateImagePhotoshootMetadata,
  )
```

Reply:
Your pathological dislike of nesting pattern matchers meets mine of _'s. I tried to pull this logic out into custom unapplies, but couldn't pull it off.